lib/ipam/ipam.go

```diff
@@ -205,42 +206,76 @@ func (c ipamClient) getBlockFromAffinity(ctx context.Context, aff *model.KVPair)
 // determinePools compares a list of requested pools with the enabled pools
 // and returns the intersect. If any requested pool does not exist, or is not enabled, an error is returned.
 // If no pools are requested, all enabled pools are returned.
-func (c ipamClient) determinePools(requestedPoolNets []net.IPNet, version int) ([]v3.IPPool, error) {
+func (c ipamClient) determinePools(requestedPoolNets []net.IPNet, version int, node *v3.Node) ([]v3.IPPool, error) {
```

Comment:
I think this function has become complicated enough that it should have its own set of unit tests to assert correct behavior. A few I think we should have:

- Default case (returns all enabled pools, filters out disabled pools)
- Pass in a few requested pools, should do the right thing.
- Don't pass in requested pools, some enabled pools have nodeselectors which match the node, some don't
- Pass in requested pools, some requested pools have nodeselectors, some don't.
- some other cases, probably.